### PR TITLE
Translation providers improvements

### PR DESCRIPTION
**Pull request by fxprunayre**
###  Better handle record which may contains languages

* Add Anchor support
* better preserve existing translations
* add translate all,
* translate only empty.
* Also experiment support in suggestion in 115-3 (disabled for now - list of String is not supported, only one xpath at a time).

image

### Configuration 

* Create a directive that can be reuse in any harvester
* Add comon xpath selector

image

### Add google translate service

image

Based on https://stackoverflow.com/questions/8147284/how-to-use-google-translate-api-in-my-java-application

### Misc

* Improve translation based on @MichelGabriel feedback https://github.com/geonetwork/core-geonetwork/pull/7849#issuecomment-2017465832

  

# Checklist

- [ ] I have read the [contribution guidelines](https://github.com/geonetwork/core-geonetwork/blob/main/CONTRIBUTING.md)
- [ ] *Pull request* provided for `main` branch, backports managed with label
- [ ] *Good housekeeping* of code, cleaning up comments, tests, and documentation
- [ ] *Clean commit history* broken into understandable chucks, avoiding big commits with hundreds of files, cautious of reformatting and whitespace changes
- [ ] *Clean commit message*s, longer verbose messages are encouraged
- [ ] *API Changes* are identified in commit messages
- [ ] *Testing* provided for features or enhancements using [automatic tests](https://github.com/geonetwork/core-geonetwork/blob/main/software_development/TESTING.md)
- [ ] *User documentation* provided for new features or enhancements in [manual](https://github.com/geonetwork/core-geonetwork/tree/main/docs/manual)
- [ ] *Build documentation* provided for development instructions in `README.md` files
- [ ] *Library management* using `pom.xml` dependency management. Update build documentation with intended library use and library tutorials or documentation

### CODE REVIEW

File: schemas/iso19115-3.2018/src/main/plugin/iso19115-3.2018/process/translate.xsl [hunk 1]
```diff
@@ -32,225 +32,176 @@
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
                 xmlns:xs="http://www.w3.org/2001/XMLSchema"
-                xmlns:gn-fn-metadata="http://geonetwork-opensource.org/xsl/functions/metadata"
+                xmlns:geonet="http://www.fao.org/geonetwork"
```
Comment (josegar74) [inline schemas/iso19115-3.2018/src/main/plugin/iso19115-3.2018/process/translate.xsl:35]: Is it required a template to remove `geonet` elements also, other xslt in the `process` folder has that type of template?

in the same xslt for iso19139 this line doesn't exist.

Comment (fxprunayre) [reply schemas/iso19115-3.2018/src/main/plugin/iso19115-3.2018/process/translate.xsl:35]: For harvesting no. I only tested suggestion with 115-3 for now but turned it off as it requires a bit more effort. Maybe we can check later when we add translation service in the editor and also align 19139 ?

Comment (fxprunayre) [reply schemas/iso19115-3.2018/src/main/plugin/iso19115-3.2018/process/translate.xsl:35]: `geonet` namespace is required for suggestion templates

File: translationproviders/src/main/java/org/fao/geonet/translations/googletranslate/GoogleTranslateService.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ */
+
+package org.fao.geonet.translations.googletranslate;
+
+import org.fao.geonet.kernel.setting.SettingManager;
+import org.fao.geonet.kernel.setting.Settings;
+import org.fao.geonet.translations.ITranslationService;
+import org.fao.geonet.translations.libretranslate.LibreTranslateClient;
+import org.fao.geonet.translations.libretranslate.LibreTranslateClientException;
+import org.fao.geonet.translations.libretranslate.LibreTranslateResponse;
```
Comment (josegar74) [inline translationproviders/src/main/java/org/fao/geonet/translations/googletranslate/GoogleTranslateService.java:31]:
```suggestion
```

File: translationproviders/src/main/java/org/fao/geonet/translations/libretranslate/LibreTranslateClientException.java [hunk 3]
```diff
@@ -1,3 +1,25 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ *
+ * Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
+ * Rome - Italy. email: geonetwork@osgeo.org
+ */
 package org.fao.geonet.translations.libretranslate;
 
 public class LibreTranslateClientException extends Exception{
```
Comment (josegar74) [inline translationproviders/src/main/java/org/fao/geonet/translations/libretranslate/LibreTranslateClientException.java:25]:
```suggestion
public class LibreTranslateClientException extends Exception {
```

File: web-ui/src/main/resources/catalog/components/admin/harvester/HarvesterDirective.js [hunk 4]
```diff
@@ -130,6 +131,82 @@
     }
   ]);
 
+  /**
+   * Display translation service configuration
+   */
+  module.directive("gnHarvesterTranslateContentConfigurator", [
```
Comment (josegar74) [inline web-ui/src/main/resources/catalog/components/admin/harvester/HarvesterDirective.js:137]: With this refactor the language picker in the harvester doesn't work. Clicking in the field to select a language:

```
UtilityDirective.js:1029 Uncaught TypeError: Cannot read properties of null (reading 'all')
    at Dataset.allOrSearchFn [as source] (UtilityDirective.js:1029:33)
    at Dataset.update (typeahead.bundle.js?v=e6c3a99c1fbe6adea82e4e57eceddf502bff2fe9:1708:22)
    at updateDataset (typeahead.bundle.js?v=e6c3a99c1fbe6adea82e4e57eceddf502bff2fe9:1889:29)
```

File: web-ui/src/main/resources/catalog/components/admin/harvester/partials/translate.html [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      class="control-label"
+      data-translate=""
+      >harvesterTranslateContentLangs</label
+    >
+    <div
+      data-gn-data-picker=""
+      data-source="languageSource"
+      data-value="content.translateContentLangs"
+      data-element-id="csw-languagepicker"
+    />
```
Comment (josegar74) [inline web-ui/src/main/resources/catalog/components/admin/harvester/partials/translate.html:33]:
```suggestion
    ></div>
```

File: web-ui/src/main/resources/catalog/components/admin/harvester/partials/translate.html [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        </ul>
+      </div>
+    </div>
+
+    <div
+      id="gn-harvest-settings-csw-advanced-translatefields-list"
+      style="height: 300px"
+      ui-ace="{useWrapMode:true, showGutter:true, mode:'text'}"
+      data-ng-model="content.translateContentFields"
+    />
```
Comment (josegar74) [inline web-ui/src/main/resources/catalog/components/admin/harvester/partials/translate.html:74]:
```suggestion
    ></div>
```

File: web-ui/src/main/resources/catalog/components/admin/harvester/HarvesterDirective.js (hunk 4 above)
Comment (josegar74) [reply web-ui/src/main/resources/catalog/components/admin/harvester/HarvesterDirective.js:137]: It works in debug mode, but not in the production mode

**Review by josegar74 — APPROVED** [review]

**Pull request merged**